### PR TITLE
Case study: E3SM-IO with Cache VOL and Async VOL

### PR DESCRIPTION
**Pull request by yzanhua**
This case study shows how to run E3SM-IO with Cache Vol and Async Vol.

### CODE REVIEW

**Comment by wkliao** [comment]
Please use VOL, all capitals.

File: README.md [hunk 1]
```diff
@@ -32,6 +32,7 @@ log-based layout in files.
 * Current build status:
   [![MPICH](https://github.com/Parallel-NetCDF/E3SM-IO/actions/workflows/mpich_static.yml/badge.svg)](https://github.com/Parallel-NetCDF/E3SM-IO/actions/workflows/mpich_static.yml)
   [![OpenMPI](https://github.com/Parallel-NetCDF/E3SM-IO/actions/workflows/ubuntu_ompi.yml/badge.svg)](https://github.com/Parallel-NetCDF/E3SM-IO/actions/workflows/ubuntu_ompi.yml)
+* Run E3SM-IO with Cache VOL and Async VOL: see [cache_async_vol.md](./cache_async_vol.md)
```
Comment (wkliao) [inline README.md:35]: This bullet should goes to docs/INSTALL.md.
Also, move cache_async_vol.md to folder docs.

**Comment by wkliao** [comment]
Please embed the URLs of Cache VOL and Async VOL when referring to them.

**Comment by wkliao** [comment]
I am seeing the below configure options when building HDF5. Are they required
for Cache VOL? If yes, please use a URL link pointing to the Cache VOL document
say so.
```
--enable-threadsafe --enable-unsupported 
```
Add the sentence in E3SM-IO build instructions about `--enable-threading`
option is required (because it is required by Async VOL or Cache VOL as well?).

Similarly we need a reference to the requirement of `Argobots, required by Async VOL`

For the commands, please do not jam them into one line.
Keep one line per command. For example, do not use `make; make install`

**Comment by wkliao** [comment]
@khou2020 
Can you check the timings in the md file?
```
Time of define variables           min/max =  58.7400 /  65.5300
Time of posting write requests     min/max =  35.6823 /  35.6953
Time of write flushing             min/max =   0.3860 /   0.3990
Time of close                      min/max =   5.0383 /   5.0620
end-to-end time                    min/max = 107.3378 / 107.3614
```
The posting and flushing are very slow. Are they normal?
This is the 16p small F case.

**Comment by yzanhua** [comment]
1. Rerun on ECP using new decomposition file datasets/map_f_case_16p.h5
2. The last two commits above (f1fcae9 and 8f00d09) contain the updated results. Other commits are due to a rebase on to master branch.

**Comment by wkliao** [comment]
Did you re-run the example? The example outputs still use the old map file name.

**Comment by yzanhua** [comment]
I did not make myself clear.
The last commit above (8f00d09) contains the updated results for the re-run of new map file name.

**Comment by wkliao** [comment]
Can you also re-run it without Cache/Async VOL?
Just to see how different it can be.

**Comment by yzanhua** [comment]
Without Cache and Async VOL (using native VOL), I got a (much) better timing result.

If I only use Cache VOL (without using Async VOL), the timing results are close to using native VOL.

Should I put these results somewhere in the case_study?

<details> <summary> Native VOL (click here to see/unsee timing results)</summary>

```
==== Benchmarking F case =============================
Total number of MPI processes      = 16
Number of IO processes             = 16
Input decomposition file           = /files2/scratch/zhd1108/E3SM-IO/E3SM-IO/datasets/map_f_case_16p.h5
Number of decompositions           = 3
Output file/directory              = can_F_out.h5
Using noncontiguous write buffer   = no
Variable write order: same as variables are defined
==== HDF5 canonical I/O ==============================
History output file                = can_F_out_h0.h5
No. variables use no decomposition =     27
No. variables use decomposition D0 =      1
No. variables use decomposition D1 =    323
No. variables use decomposition D2 =     63
Total no. climate variables        =    414
Total no. attributes               =   1421
Total no. noncontiguous requests   = 4207968
Max   no. noncontiguous requests   = 272135
Min   no. noncontiguous requests   = 252670
Write no. records (time dim)       =      1
I/O flush frequency                =      1
No. I/O flush calls                =      1
-----------------------------------------------------------
Total write amount                         = 16.97 MiB = 0.02 GiB
Time of I/O preparing              min/max =   0.0010 /   0.0014
Time of file open/create           min/max =   0.0017 /   0.0019
Time of define variables           min/max =   0.0452 /   0.0452
Time of posting write requests     min/max =   3.8051 /   3.8052
Time of write flushing             min/max =   0.0000 /   0.0000
Time of close                      min/max =   0.0233 /   0.0234
end-to-end time                    min/max =   3.8771 /   3.8772
Emulate computation time (sleep)   min/max =   2.0000 /   2.0000
I/O bandwidth in MiB/sec (write-only)      = 4.4592
I/O bandwidth in MiB/sec (open-to-close)   = 4.3764
-----------------------------------------------------------
==== Benchmarking F case =============================
Total number of MPI processes      = 16
Number of IO processes             = 16
Input decomposition file           = /files2/scratch/zhd1108/E3SM-IO/E3SM-IO/datasets/map_f_case_16p.h5
Number of decompositions           = 3
Output file/directory              = can_F_out.h5
Using noncontiguous write buffer   = no
Variable write order: same as variables are defined
==== HDF5 canonical I/O ==============================
History output file                = can_F_out_h1.h5
No. variables use no decomposition =     27
No. variables use decomposition D0 =      1
No. variables use decomposition D1 =     22
No. variables use decomposition D2 =      1
Total no. climate variables        =     51
Total no. attributes               =    142
Total no. noncontiguous requests   = 400226
Max   no. noncontiguous requests   =  25963
Min   no. noncontiguous requests   =  24026
Write no. records (time dim)       =      5
I/O flush frequency                =      1
No. I/O flush calls                =      5
-----------------------------------------------------------
Total write amount                         = 1.88 MiB = 0.00 GiB
Time of I/O preparing              min/max =   0.0000 /   0.0000
Time of file open/create           min/max =   0.0019 /   0.0019
Time of define variables           min/max =   0.0053 /   0.0053
Time of posting write requests     min/max =  10.1859 /  10.1865
Time of write flushing             min/max =   0.0000 /   0.0000
Time of close                      min/max =   0.0023 /   0.0023
end-to-end time                    min/max =  10.1961 /  10.1961
Emulate computation time (sleep)   min/max =   2.0000 /   2.0000
I/O bandwidth in MiB/sec (write-only)      = 0.1843
I/O bandwidth in MiB/sec (open-to-close)   = 0.1841
-----------------------------------------------------------
read_decomp=0.01 e3sm_io_core=14.07 MPI init-to-finalize=14.08
-----------------------------------------------------------
```

</details>

<details> <summary> Cache VOL only (click here to see/unsee timing results)</summary>

```
==== Benchmarking F case =============================
Total number of MPI processes      = 16
Number of IO processes             = 16
Input decomposition file           = /files2/scratch/zhd1108/E3SM-IO/E3SM-IO/datasets/map_f_case_16p.h5
Number of decompositions           = 3
Output file/directory              = can_F_out.h5
Using noncontiguous write buffer   = no
Variable write order: same as variables are defined
==== HDF5 canonical I/O ==============================
History output file                = can_F_out_h0.h5
No. variables use no decomposition =     27
No. variables use decomposition D0 =      1
No. variables use decomposition D1 =    323
No. variables use decomposition D2 =     63
Total no. climate variables        =    414
Total no. attributes               =   1421
Total no. noncontiguous requests   = 4207968
Max   no. noncontiguous requests   = 272135
Min   no. noncontiguous requests   = 252670
Write no. records (time dim)       =      1
I/O flush frequency                =      1
No. I/O flush calls                =      1
-----------------------------------------------------------
Total write amount                         = 16.97 MiB = 0.02 GiB
Time of I/O preparing              min/max =   0.0006 /   0.0012
Time of file open/create           min/max =   0.0022 /   0.0024
Time of define variables           min/max =   0.0585 /   0.0586
Time of posting write requests     min/max =   3.9087 /   3.9089
Time of write flushing             min/max =   0.0000 /   0.0000
Time of close                      min/max =   0.0344 /   0.0344
end-to-end time                    min/max =   4.0055 /   4.0055
Emulate computation time (sleep)   min/max =   2.0000 /   2.0000
I/O bandwidth in MiB/sec (write-only)      = 4.3409
I/O bandwidth in MiB/sec (open-to-close)   = 4.2362
-----------------------------------------------------------
==== Benchmarking F case =============================
Total number of MPI processes      = 16
Number of IO processes             = 16
Input decomposition file           = /files2/scratch/zhd1108/E3SM-IO/E3SM-IO/datasets/map_f_case_16p.h5
Number of decompositions           = 3
Output file/directory              = can_F_out.h5
Using noncontiguous write buffer   = no
Variable write order: same as variables are defined
==== HDF5 canonical I/O ==============================
History output file                = can_F_out_h1.h5
No. variables use no decomposition =     27
No. variables use decomposition D0 =      1
No. variables use decomposition D1 =     22
No. variables use decomposition D2 =      1
Total no. climate variables        =     51
Total no. attributes               =    142
Total no. noncontiguous requests   = 400226
Max   no. noncontiguous requests   =  25963
Min   no. noncontiguous requests   =  24026
Write no. records (time dim)       =      5
I/O flush frequency                =      1
No. I/O flush calls                =      5
-----------------------------------------------------------
Total write amount                         = 1.88 MiB = 0.00 GiB
Time of I/O preparing              min/max =   0.0000 /   0.0000
Time of file open/create           min/max =   0.0024 /   0.0024
Time of define variables           min/max =   0.0078 /   0.0078
Time of posting write requests     min/max =  10.1861 /  10.1867
Time of write flushing             min/max =   0.0000 /   0.0000
Time of close                      min/max =   0.0024 /   0.0028
end-to-end time                    min/max =  10.1995 /  10.1998
Emulate computation time (sleep)   min/max =   2.0000 /   2.0000
I/O bandwidth in MiB/sec (write-only)      = 0.1843
I/O bandwidth in MiB/sec (open-to-close)   = 0.1841
-----------------------------------------------------------
read_decomp=0.01 e3sm_io_core=14.21 MPI init-to-finalize=14.21
-----------------------------------------------------------
```

</details>

**Comment by khou2020** [comment]
> @khou2020 Can you check the timings in the md file?
> 
> ```
> Time of define variables           min/max =  58.7400 /  65.5300
> Time of posting write requests     min/max =  35.6823 /  35.6953
> Time of write flushing             min/max =   0.3860 /   0.3990
> Time of close                      min/max =   5.0383 /   5.0620
> end-to-end time                    min/max = 107.3378 / 107.3614
> ```
> 
> The posting and flushing are very slow. Are they normal? This is the 16p small F case.

It depends on which machine it was ran on.
I don't know about define variables. Using native VOL shouldn't take that long.
Post can be long if you call one H5Dwrite per non-contig request.

**Comment by wkliao** [comment]
We no longer call H5Dwrite one per noncontiguous request.

**Comment by wkliao** [comment]
Please add a new github action CI for testing the use of Cache and Async VOLs.

**Pull request merged**